source/extensions/filters/http/grpc_web/grpc_web_filter.cc

```diff
@@ -164,6 +194,15 @@ Http::FilterDataStatus GrpcWebFilter::encodeData(Buffer::Instance& data, bool) {
     return Http::FilterDataStatus::Continue;
   }
 
+  // When the upstream response is not a gRPC response, we set the "grpc-message" header with the
+  // upstream body content.
+  if (response_headers_ != nullptr) {
+    data.drain(data.length());
+    response_headers_->setGrpcMessage(Http::Utility::PercentEncoding::encode(data.toString()));
+    response_headers_->setContentLength(0);
+    return Http::FilterDataStatus::Continue;
+  }
```

Comment:
I'm not sure if this works right if there are multiple data frames? It also might not be feasible if the upstream body is huge

Reply:
Yeah, will update this.

Reply:
Updated with no buffering if we know the response is wrong, and take the last frame (when end_stream true) as the grpc-message content. Let me know what you think.